File: public/static/templates/go/common/KafkaProcessor.go
forEach: BoundedContext
fileName: KafkaProcessor.go
path: {{name}}/{{name}}
---
package {{name}}

import (
	"gopkg.in/confluentinc/confluent-kafka-go.v1/kafka"
	"fmt"
    "encoding/json"
    "{{name}}/config"
)

var producer *kafka.Producer
var consumer *kafka.Consumer
var topic string

func InitProducer(platform string){
	var err error
    var options = config.Reader(platform)
	producer, err = kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": options["bootstrap_servers"]})
	if err != nil {
		panic(err)
	}
	topic = options["destination"]
}

func InitConsumer(platform string){
	var err error
    var options = config.Reader(platform)
	consumer, err = kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": options["bootstrap_servers"],
		"group.id":          options["group_id"],
		"broker.address.family": "v4",
		"session.timeout.ms":    6000,
		"auto.offset.reset": "earliest",
	})
	topic = options["destination"]

	if err != nil {
		panic(err)
	}
	defer consumer.Close()
	KafkaConsumer()
	
}

func KafkaProducer() (*kafka.Producer, string){
	
	return producer, topic
}

func KafkaConsumer(){
    
	
    consumer.SubscribeTopics([]string{topic}, nil)

	var dat map[string]interface{}
    for {
        msg, err := consumer.ReadMessage(-1)
        if err == nil {
			if err := json.Unmarshal(msg.Value, &dat); err != nil {
				panic(err)
			}
            {{#policies}}
				{{#relationEventInfo}}
			for _, header := range msg.Headers {
				if header.Key == "type" && string(header.Value) == "{{eventValue.namePascalCase}}" {
					whenever{{eventValue.namePascalCase}}_{{../namePascalCase}}(dat)
				}
			}
            //if dat["eventType"] == "{{eventValue.namePascalCase}}"{
            //    whenever{{eventValue.namePascalCase}}_{{../namePascalCase}}(dat)
            //}
				{{/relationEventInfo}}
            {{/policies}}
			{{#views}}
				{{#createRules}}
			for _, header := range msg.Headers {
				if header.Key == "type" && string(header.Value) == "{{when.namePascalCase}}" {
					when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
				}
			}
			//if dat["eventType"] == "{{when.namePascalCase}}"{
			//	when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
			//}
				{{/createRules}}
				{{#updateRules}}
			for _, header := range msg.Headers {
				if header.Key == "type" && string(header.Value) == "{{when.namePascalCase}}" {
					when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
				}
			}
			//if dat["eventType"] == "{{when.namePascalCase}}"{
			//	when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
			//}
				{{/updateRules}}

				{{#deleteRules}}
			for _, header := range msg.Headers {
				if header.Key == "type" && string(header.Value) == "{{when.namePascalCase}}" {
					when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
				}
			}
			//if dat["eventType"] == "{{when.namePascalCase}}"{
			//	when{{when.namePascalCase}}_then_{{operation}}_{{_index}}(dat)
			//}
				{{/deleteRules}}
			{{/views}}
			
        } else {
            // The client will automatically try to recover from all errors.
            fmt.Printf("Consumer error: %v (%v)\n", err, msg)
        }
    }
}

func SendStreamMsg(message, eventType string) {
	producer, topic := KafkaProducer()
	eventHeaders := []kafka.Header{
		{"type", []byte(eventType)},
	}
	producer.Produce(&kafka.Message{
		TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
		Value:          []byte(message),
		Headers:        eventHeaders,
	}, nil)
}

File: public/static/templates/python/KafkaProcessor.py
forEach: BoundedContext
fileName: KafkaProcessor.py
path: {{name}}
---
from kafka import KafkaProducer
from flask_kafka import FlaskKafka
from threading import Event
import json
from config import config_reader
{{#policies}}
{{#relationEventInfo}}
from PolicyHandler import whenever{{eventValue.namePascalCase}}_{{../namePascalCase}}
{{/relationEventInfo}}
{{/policies}}

class StreamHandler():
	
	def __init__(self):
		INTERRUPT_EVENT = Event()
		config = config_reader.reader()
		self.destination = config['destination']
		self.group_id = config['group_id']
		self.bootstrap_server = config['bootstrap_servers']
		self.producer = KafkaProducer(acks=0, compression_type='gzip', bootstrap_servers=[self.bootsrap_server], value_serializer=lambda x: json.dumps(x).encode('utf-8'))
		self.consumer = FlaskKafka(INTERRUPT_EVENT,
		         bootstrap_servers=[self.bootstrap_server],
		         group_id= self.group_id
		         )


	def produce(self, msg):
		self.producer.send(self.destination, value=msg)

streamhandler = StreamHandler()

@streamhandler.consumer.handle(streamhandler.destination)
def consume(msg):
    my_json = msg.value.decode('utf8').replace("'", '"')
    data = json.loads(my_json)
    {{#policies}}
    if data['eventType'] == {{#relationEventInfo}}"{{eventValue.namePascalCase}}"{{/relationEventInfo}}:
        {{#relationEventInfo}}
        whenever{{eventValue.namePascalCase}}_{{../namePascalCase}}(data)
        {{/relationEventInfo}}
    {{/policies}}